homeassistant/components/subaru/__init__.py

```diff
@@ -0,0 +1,192 @@
+        )
+        _LOGGER.debug("Using subarulink %s", controller.version)
+        await controller.connect()
+    except InvalidCredentials:
+        _LOGGER.error("Invalid account")
+        return False
+    except SubaruException as err:
+        raise ConfigEntryNotReady(err.message) from err
+
+    vehicle_info = {}
+    for vin in controller.get_vehicles():
+        vehicle_info[vin] = get_vehicle_info(controller, vin)
+
+    async def async_update_data():
+        """Fetch data from API endpoint."""
+        try:
+            return await refresh_subaru_data(entry, vehicle_info, controller)
+        except SubaruException as err:
+            raise UpdateFailed(err.message) from err
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name=COORDINATOR_NAME,
+        update_method=async_update_data,
+        update_interval=timedelta(seconds=FETCH_INTERVAL),
+    )
+
+    await coordinator.async_refresh()
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        ENTRY_CONTROLLER: controller,
+        ENTRY_COORDINATOR: coordinator,
+        ENTRY_VEHICLES: vehicle_info,
+    }
+
+    for component in SUPPORTED_PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    async def async_fetch_service(call):
+        """Force refresh of data from Subaru."""
+        vin = call.data[VEHICLE_VIN].upper()
+        if vin not in vehicle_info.keys():
+            hass.components.persistent_notification.create(
+                f"ERROR - Invalid VIN provided while calling {call.service}", "Subaru"
+            )
+        else:
+            if call.service == REMOTE_SERVICE_FETCH:
+                await coordinator.async_refresh()
+
+    hass.services.async_register(
+        DOMAIN, REMOTE_SERVICE_FETCH, async_fetch_service, schema=REMOTE_SERVICE_SCHEMA
+    )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in SUPPORTED_PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+    return unload_ok
+
+
+async def refresh_subaru_data(config_entry, vehicle_info, controller):
+    """
+    Refresh local data with data fetched via Subaru API.
+
+    Subaru API calls assume a server side vehicle context
+    Data fetch/update must be done for each vehicle
+    """
+    data = {}
+
+    for vehicle in vehicle_info.values():
+        vin = vehicle[VEHICLE_VIN]
+
+        # Active subscription required
+        if not vehicle[VEHICLE_HAS_SAFETY_SERVICE]:
+            continue
+
+        # Optionally send an "update" remote command to vehicle (throttled with update_interval)
+        if config_entry.options.get(CONF_UPDATE_ENABLED, False):
```

Comment:
If the api can handle multiple requests at once make the inner block a coroutine and use asyncio.gather to call them in parallel

Reply:
Unfortunately the api can only handle one request at a time.